src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,154 @@
+                         NativeCRef<OpticalPropertyData> const& properties,
+                         NativeCRef<CerenkovData> const& shared,
+                         OpticalPreGenStepData const& step_data);
+
+    // Return a populated optical distribution data for the Cerenkov Generator
+    template<class Generator>
+    inline CELER_FUNCTION OpticalDistributionData operator()(Generator& rng);
+
+  private:
+    units::ElementaryCharge charge_;
+    real_type step_len_;
+    OpticalMaterialId mat_id_;
+    OpticalPreGenStepData step_data_;
+    real_type num_photons_per_len_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with optical properties, Cerenkov, and step information.
+ */
+CELER_FUNCTION CerenkovPreGenerator::CerenkovPreGenerator(
+    ParticleTrackView const& particle_view,
+    SimTrackView const& sim_view,
+    OpticalMaterialId mat_id,
+    NativeCRef<OpticalPropertyData> const& properties,
+    NativeCRef<CerenkovData> const& shared,
+    OpticalPreGenStepData const& step_data)
+    : charge_(particle_view.charge())
+    , step_len_(sim_view.step_length())
+    , mat_id_(mat_id)
+    , step_data_(step_data)
+{
+    CELER_EXPECT(charge_ != zero_quantity());
+    CELER_EXPECT(step_len_ > 0);
+    CELER_EXPECT(mat_id_);
+    CELER_EXPECT(step_data_);
+
+    auto const& pre = step_data_.points[StepPoint::pre];
+    auto const& post = step_data_.points[StepPoint::post];
+    units::LightSpeed beta(real_type{0.5}
+                           * (pre.speed.value() + post.speed.value()));
+
+    CerenkovDndxCalculator calculate_dndx(properties, shared, mat_id_, charge_);
+    num_photons_per_len_ = calculate_dndx(beta);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return an \c OpticalDistributionData object. If no photons are sampled, an
+ * empty object is returned and can be verified via its own operator bool.
+ *
+ * The number of photons is sampled from a Poisson distribution with a mean
+ * \f[
+   \langle n \rangle = \ell_\text{step} \frac{dN}{dx}
+ * \f]
+ * where \f$ \ell_\text{step} \f$ is the step length.
+ */
+template<class Generator>
+CELER_FUNCTION OpticalDistributionData
+CerenkovPreGenerator::operator()(Generator& rng)
+{
+    if (num_photons_per_len_ == 0)
+    {
+        return {};
+    }
+
+    // Sample number of photons from a Poisson distribution
+    auto sampled_num_photons = PoissonDistribution<real_type>(
+        num_photons_per_len_ * step_len_)(rng);
+
+    OpticalDistributionData data;
+    data.num_photons = sampled_num_photons;
+    data.time = step_data_.time;
+    data.step_length = step_len_;
+    data.charge = charge_;
+    data.material = mat_id_;
+    data.points = step_data_.points;
```

Comment:
I guess we could. But we'd need an `OpticalDistributionData result_;` private member. I don't think it makes much difference, since we always construct and operate once. Maybe it's bit better as is, since we only spend time copying data if `operator()` is successful.

Reply:
No we *definitely* want to return a sample result...

Reply:
I mean, we still return a sample, and it is still false if num_photons is zero. So copying at construction and `return result_` in `operator()` seems like it would behave identically. We just moved the copy assignment to the construction (which when num_photons is zero, becomes unnecessary).